In [ ]:
import pandas as pd
import numpy as np
import datetime as dt


# Библиотека pandas

## Занятие 1.

Тема занятия: библиотека `pandas`: знакомство, предназначение, связь с другими библиотеками

Рассматриваемые вопросы:

1. История развития библиотеки `pandas`; 
2. Предназначение библиотеки `pandas`;
3. Совместное использование с другими библиотеками;
4. Установка библиотеки `pandas`;
5. Объекты для работы с данными;
6. Загрузка данных;
7. Знакомство с данными. 

### История развития библиотеки `pandas`
***
<u>Замечание (этимология).</u> Название `pandas` образовано как от `panel data` (панельные данные), применяемого в эконометрике термина для обозначения многомерных структурированных наборов данных, так и от фразы `Python data analysis` [1].
***
<figure>
<center>
<img src='https://drive.google.com/uc?export=view&id=17c-ZeakLjKOxWepNCS0wpXV-oeakPBb9' width="500" height="300" />
<figcaption>Больша́я па́нда или бамбу́ковый медведь</figcaption></center>
</figure>

`Pandas` --- это программная библиотека на языке Python, которая предоставляет очень удобные, с точки зрения использования, инструменты для хранения и работы с данными.

Разработку библиотеки `pandas` начал Уэс Маккини в 2008 г. К концу 2009 года она представлена как проект с открытым программным кодом. С 2015 года `pandas` является спонсируемым проектом [NumFOCUS](https://numfocus.org/sponsored-projects). В настоящее время библиотека активно курируется и разрабатывается различными организациями и участниками. 

***

<u>Замечание (для общего развития).</u>
NumFOCUS - это общественная благотворительная организация в США.

Миссия NumFOCUS - продвигать открытые практики в исследованиях, данных и научных вычислениях, выступая в качестве финансового спонсора проектов с открытым исходным кодом и организуя образовательные программы, ориентированные на сообщества.
***

Официальная страница библиотеки находится по адресу [https://pandas.pydata.org/](https://pandas.pydata.org/).

### Предназначение библиотеки `pandas`
<figure>
<center>
<img src='https://drive.google.com/uc?export=view&id=1mHdn3yjiz233ANaXOcdo5WpddkNBWAHN'/>
<figcaption>Процесс анализа данных</figcaption></center>
</figure>

### Совместное использование с другими библиотеками

Библиотека `pandas` довольно часто используется совместно со следующими библиотеками:
- численные и научные вычисления: [`numpy`](https://numpy.org/), [`scipy`](https://scipy.org/);
- статистический анализ: [`statsmodels`](https://www.statsmodels.org/);
- машинное обучение: [`scikit-learn`](https://scikit-learn.org);
- визуализация: [`matplotlib`](https://matplotlib.org/), [`seaborn`](https://seaborn.pydata.org/).




### Установка библиотеки `pandas`

При изложении этого вопроса будем предполагать, что на компьютере 
- установлен [`Anaconda`](https://www.anaconda.com/products/individual) $-$ дистрибутив языка программирования `Python`;
- актуальная версия одного из современных веб-браузеров (Google Chrome, Internet Explorer, Mozilla Firefox и т.п.).

И есть возможность работать с либо [Jupyter Notebook](https://jupyter.org/), либо [Colab](https://colab.research.google.com/).

***
<u>Замечание (установка).</u> Об особенностях установки дистрибутива [`Anaconda`](https://www.anaconda.com/products/individual) под различные операционные системы можно почитать в работе [2]. 
***

Далее будем предполагать, что работа ведется либо в [Jupyter Notebook](https://jupyter.org/), либо в [Colab](https://colab.research.google.com/).

Рассмотрим следующие вопросы, связанные с установкой библиотеки `pandas`:
- наличие библиотеки;
- установка библиотеки;
- проверка обновлений;
- установка библиотеки требуемой версии;
- удаление библиотеки;
- подключение библиотеки к проекту;
- работа с файлом требований;
- установка параметров вывода библиотеки.     

#### Наличие библиотеки

Для того, чтобы проверить наличие библиотеки `pandas` в системе следует воспользоваться одним из следующих вариантов команд:
- `pip show pandas` возвращает версию библиотеки, ее местонахождение на диске и зависимости;
- `pip list` --- возвращает установленные библиотеки вместе с их версиями;
- `pip freeze` --- возвращает установленные библиотеки вместе с их версиями (упомянуть про файл с требованиями).

Полезная команда `pd.show_versions()`

In [1]:
pip show pandas

Name: pandas
Version: 1.3.5
Summary: Powerful data structures for data analysis, time series, and statistics
Home-page: https://pandas.pydata.org
Author: The Pandas Development Team
Author-email: pandas-dev@python.org
License: BSD-3-Clause
Location: /usr/local/lib/python3.7/dist-packages
Requires: numpy, pytz, python-dateutil
Required-by: xarray, vega-datasets, statsmodels, sklearn-pandas, seaborn, pymc, prophet, plotnine, pandas-profiling, pandas-gbq, pandas-datareader, mlxtend, mizani, holoviews, gspread-dataframe, google-colab, fix-yahoo-finance, fastai, db-dtypes, datascience, cufflinks, cmdstanpy, arviz, altair


In [ ]:
pip show pandas

Name: pandas
Version: 1.3.5
Summary: Powerful data structures for data analysis, time series, and statistics
Home-page: https://pandas.pydata.org
Author: The Pandas Development Team
Author-email: pandas-dev@python.org
License: BSD-3-Clause
Location: /usr/local/lib/python3.7/dist-packages
Requires: numpy, pytz, python-dateutil
Required-by: xarray, visions, vega-datasets, statsmodels, sklearn-pandas, seaborn, pymc3, plotnine, phik, pandas-profiling, pandas-gbq, pandas-datareader, mlxtend, mizani, holoviews, gspread-dataframe, google-colab, fix-yahoo-finance, fbprophet, fastai, cufflinks, cmdstanpy, arviz, altair


In [ ]:
pd.__version__

'1.3.5'

In [ ]:
print(pd.show_versions())

/usr/local/lib/python3.7/dist-packages/psycopg2/__init__.py:144: UserWarning: The psycopg2 wheel package will be renamed from release 2.8; in order to keep installing from binary please use "pip install psycopg2-binary" instead. For details see: <http://initd.org/psycopg/docs/install.html#binary-install-from-pypi>.
  """)



INSTALLED VERSIONS
------------------
commit           : 66e3805b8cabe977f40c05259cc3fcf7ead5687d
python           : 3.7.13.final.0
python-bits      : 64
OS               : Linux
OS-release       : 5.4.188+
Version          : #1 SMP Sun Apr 24 10:03:06 PDT 2022
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 1.3.5
numpy            : 1.21.6
pytz             : 2022.1
dateutil         : 2.8.2
pip              : 21.1.3
setuptools       : 57.4.0
Cython           : 0.29.30
pytest           : 3.6.4
hypothesis       : None
sphinx           : 1.8.6
blosc            : None
feather          : 0.4.1
xlsxwriter       : None
lxml.etree       : 4.2.6
html5lib         : 1.0.1
pymysql          : None
psycopg2         : 2.7.6.1 (dt dec pq3 ext lo64)
jinja2           : 2.11.3
IPython          : 5.5.0
pandas_datareader: 0.9.0
bs4              : 4.6.3
bottleneck       : 1

In [ ]:
pip list

Package                       Version
----------------------------- --------------
absl-py                       0.12.0
alabaster                     0.7.12
albumentations                0.1.12
altair                        4.1.0
appdirs                       1.4.4
argcomplete                   1.12.3
argon2-cffi                   21.1.0
arviz                         0.11.4
astor                         0.8.1
astropy                       4.3.1
astunparse                    1.6.3
atari-py                      0.2.9
atomicwrites                  1.4.0
attrs                         21.2.0
audioread                     2.1.9
autograd                      1.3
Babel                         2.9.1
backcall                      0.2.0
beautifulsoup4                4.6.3
bleach                        4.1.0
blis                          0.4.1
bokeh                         2.3.3
Bottleneck                    1.3.2
branca                        0.4.2
bs4                           0.0.1
CacheControl

In [ ]:
pip freeze

absl-py==0.12.0
alabaster==0.7.12
albumentations==0.1.12
altair==4.1.0
appdirs==1.4.4
argcomplete==1.12.3
argon2-cffi==21.1.0
arviz==0.11.4
astor==0.8.1
astropy==4.3.1
astunparse==1.6.3
atari-py==0.2.9
atomicwrites==1.4.0
attrs==21.2.0
audioread==2.1.9
autograd==1.3
Babel==2.9.1
backcall==0.2.0
beautifulsoup4==4.6.3
bleach==4.1.0
blis==0.4.1
bokeh==2.3.3
Bottleneck==1.3.2
branca==0.4.2
bs4==0.0.1
CacheControl==0.12.10
cached-property==1.5.2
cachetools==4.2.4
catalogue==1.0.0
certifi==2021.10.8
cffi==1.15.0
cftime==1.5.1.1
chardet==3.0.4
charset-normalizer==2.0.8
click==7.1.2
cloudpickle==1.3.0
cmake==3.12.0
cmdstanpy==0.9.5
colorcet==2.0.6
colorlover==0.3.0
community==1.0.0b1
contextlib2==0.5.5
convertdate==2.3.2
coverage==3.7.1
coveralls==0.5
crcmod==1.7
cufflinks==0.17.3
cvxopt==1.2.7
cvxpy==1.0.31
cycler==0.11.0
cymem==2.0.6
Cython==0.29.24
daft==0.0.4
dask==2.12.0
datascience==0.10.6
debugpy==1.0.0
decorator==4.4.2
defusedxml==0.7.1
descartes==1.1.0
dill==0.3.4
distributed==1.25.3


#### Установка библиотеки

Для установки библиотеки `pandas` следует воспользоваться следующей командой `pip install pandas`.

In [ ]:
# pip install pandas 

#### Проверка обновлений

Команда `pip list --outdated` возвращает список устаревших библиотек.

In [ ]:
pip list --outdated

Package                       Version        Latest       Type
----------------------------- -------------- ------------ -----
absl-py                       0.12.0         1.0.0        wheel
albumentations                0.1.12         1.1.0        wheel
altair                        4.1.0          4.2.0        wheel
argon2-cffi                   21.1.0         21.3.0       wheel
attrs                         21.2.0         21.4.0       wheel
beautifulsoup4                4.6.3          4.10.0       wheel
blis                          0.4.1          0.7.5        wheel
bokeh                         2.3.3          2.4.2        wheel
cachetools                    4.2.4          5.0.0        wheel
catalogue                     1.0.0          2.0.6        wheel
chardet                       3.0.4          4.0.0        wheel
charset-normalizer            2.0.8          2.0.9        wheel
click                         7.1.2          8.0.3        wheel
cloudpickle                   1.3.0      

#### Установка библиотеки требуемой версии

Для установки библиотеки требуемой версии следует воспользоваться командой 
`pip install pandas==1.3.5`


In [ ]:
#pip install pandas==1.3.5

#### Удаление библиотеки

Для удаления любой библиотеки следует использовать команду `pip uninstall`. В частности, библиотека `pandas` удаляется следующим образом: 
```python
pip uninstall pandas
```




#### Подключение библиотеки к проекту

Для подключения библиотеки к проекту используется инструкция `import`. Она может использоваться в следующих вариантах:
- по имени библиотеки;
- по псевдониму библиотеки.

$\blacktriangleright$ По имени библиотеки:
```python
import pandas
```

$\blacktriangleright$ По псевдониму библиотеки (предпочтительный вариант):
```python
import pandas as pd
```
***

После импортирования библиотеки ее имя становится идентификатором, через который можно получить доступ к атрибутам или методам, определенным внутри библиотеки. Доступ к элементам библиотеки осуществляется с помощью точечной нотации.

<u> Пример </u>
```python
pandas.read_csv(...)
```
или 

```python
pd.read_csv(...)
```

*** 

В некоторых случаях может использоваться инструкция `from`. Ее формат таков:

```python
from <название модуля/библиотеки> import <Идентификатор 1> [as <Псевдоним 1>], 
                                  ....,  <Идентификатор N> [as <Псевдоним N>]
```
или 

```python
from <название модуля/библиотеки> import *
```
<u> Пример (несколько атрибутов) </u>
```python
from math import (pi, floor,
                  sin, cos)
```

<u> Пример (неудачное использование) </u>
```python
from pandas import read_csv as r_csv
df = r_csv('real_estate_data.csv', sep='\t')
```



#### Работа с файлом требований


Файлы требований позволяют точно указать, какие версии библиотек должны быть установлены. Команда `freeze` выводит установленные пакеты в формате требований. Можно использовать эту команду, перенаправляя вывод в файл для генерации файла требований:
```python
pip freeze >; requirements.txt
```

Для установки всех пакетов требуемых версий в другой системе, можно запустить `pip install`, указав файл требований с помощью ключа `-r`:
```python
pip install -r requirements.txt
```




##### Опции настройки вывода библиотеки `pandas`

Полный список опций представлен [здесь](https://pandas.pydata.org/pandas-docs/stable/user_guide/options.html). Дополнительную информацию можно найти в работе [3].

Для работы с настройками в библиотеке `pandas` имеются следующие функции.

|||
|---:|:---|
|Функция|Описание|
|`get_option()`|Получение значения параметра|
|`set_option()`|Установка нового значения параметра|
|`reset_option()`|Сброс параметра на значение “по умолчанию”|
|`describe_option()`|Вывод текстового описания параметра|
|`option_context()`|Присвоение параметрам новых значений в рамках определенного блока кода|

<u>Примеры использования.</u> 

$\blacktriangleright$ Для получения значения параметра используется функция `get_option()`, в качестве аргумента ей передается настраевыемый параметр в текстовом виде. 

In [ ]:
#Максимальное количество выводимых строк:
pd.get_option('display.max_rows')

60

In [ ]:
#Используемая кодировка:
pd.get_option('display.encoding')

'UTF-8'

$\blacktriangleright$ Установка нового значения осуществляется с помощью функции
`set_option()`: ее первым аргументом является название параметра в
текстовом виде, вторым - новое значение параметра.

In [ ]:
pd.get_option('display.max_rows')

5

In [ ]:
pd.set_option('display.max_rows', 5)
pd.get_option('display.max_rows')


5

$\blacktriangleright$ Значение параметра можно сбросить на значение “по умолчанию”, для этого используется функция `reset_option()`.

In [ ]:
pd.reset_option('display.max_rows')
pd.get_option('display.max_rows')


60

$\blacktriangleright$ Описание параметра можно получить с помощью функции
`describe_option()`.

In [ ]:
pd.describe_option('display.max_rows')

display.max_rows : int
    If max_rows is exceeded, switch to truncate view. Depending on
    `large_repr`, objects are either centrally truncated or printed as
    a summary view. 'None' value means unlimited.

    In case python/IPython is running in a terminal and `large_repr`
    equals 'truncate' this can be set to 0 and pandas will auto-detect
    the height of the terminal and print a truncated object which fits
    the screen height. The IPython notebook, IPython qtconsole, or
    IDLE do not run in a terminal and hence it is not possible to do
    correct auto-detection.
    [default: 60] [currently: 60]


$\blacktriangleright$ Присвоение параметрам новых значений в рамках определенного блока кода с помощью функции `option_context()`.

In [ ]:
dollars = [1, 2.09, 0.8765, 34.5]
ser_example = pd.Series(data=dollars)
ser_example

0     1.0000
1     2.0900
2     0.8765
3    34.5000
dtype: float64

In [ ]:
with pd.option_context('display.float_format', '${:,.2f}'.format):
  display(ser_example) 

0    $1.00
1    $2.09
2    $0.88
3   $34.50
dtype: float64

In [ ]:
ser_example

0     1.0000
1     2.0900
2     0.8765
3    34.5000
dtype: float64

<u>Замечание</u> Возможна работа с настройками как с атрибутами (через точку). 

In [ ]:
pd.options.display.max_rows

60

In [ ]:
pd.options.display.max_rows = 5
pd.options.display.max_rows

5

В таблице перечислены некоторые параметры, определяющие вид объектов библиотеки `pandas` и их значения по умолчанию:

|Параметр|Значение по умолчанию|Описание|
|--:|:--|:--|
|`display.encoding`|`UTF-8`|Определяет кодировку для строк, возвращаемых через метод to_string, а также для отображения строк в консоли|
|`display.max_columns`|20|Определяет максимальное количество столбцов, которые будут выведены при отображении структур `pandas`|
|`display.max_colwidth`|50|Максимальная ширина столбца (в символах) для представления структур данных `pandas`|
|`display.max_rows`|60|Определяет максимальное количество строк, которые будут выведены при отображении структур `pandas`|
|`display.precision`|6|Определяет количество знаков после запятой при выводе чисел с плавающей точкой|
|`display.float_format`|`None`|Устанавливает формат вывода чисел с плавающей точкой|




### Структуры данных в `pandas`: `Series` и `DataFrame`

#### Маркированная одномерная структура данных `Series`

`Series` -- одномерный похожий на массив объект, содержащий массив одного и того же типа данных (любого типа, поддерживаемого `numpy`) и ассоциированный с ним массив меток, который называется *индексом*. Объект `Series` часто используется для вычисления статистик по отдельному признаку. С помощью индекса может осуществляться отбок значений объекта `Series`.

Темы текущего пункта:
- создание объекта `Series`;
- работа с элементами `Series`.

##### $\blacktriangleright$ Создание объекта `Series`: общие сведения

Создать структуру `Series` можно на базе следующих типов данных:
- словарь (`dict`) `Python`;
- список (`list`) `Python`;
- массив `ndarray` (функций библиотеки `numpy`);
- скалярная величина.

Конструктор класса `Series` выглядит следующим образом:
```python
Series(data=None, index=None, dtype=None, name=None, copy=False, fastpath=False)
```
Описание параметров:
- `data`: массив, словарь, скалярное значение; значение по умолчанию -- `None`. Структура, на базе которой будет построен Series;
- `index`: одномерный массив; значение по умолчанию -- `None`. Список меток, который будет использоваться для доступа к элементам `Series`. Длина списка должна быть равна длине `data`;
- `dtype:numpy.dtype`; значение по умолчанию -- `None`. Объект, определяющий тип данных;
- `name:object`; значение по умолчанию -- `None`. Параметр, позволяющий дать имя объекту `Series` объекту; 
- `copy: bool`; значение по умолчанию -- `False`. Если параметр равен `True`, то будет создана копия массива данных;
- `fastpath: bool`; значение по умолчанию -- `False`. Внутренний параметр, [умный пользователь не должен его трогать](https://github.com/pandas-dev/pandas/issues/6903).

In [3]:
import pandas as pd
import numpy as np

In [29]:
example_series = pd.Series([10, 11.0, '12', 13, 14],
                           index = list('Январ'),
                           #dtype = 'int64',
                           name = '11111f11111irst', 
                        #    copy=False, 
                        #    fastpath=False
                           )
example_series

Я      10
н    11.0
в      12
а      13
р      14
Name: 11111f11111irst, dtype: object

In [ ]:
Январь - 10
Февраль - 11
Март - 12
Апрель - 13
Май - 14



In [ ]:
series_example['1']

-1

In [ ]:
series_example.loc['4']

1

In [ ]:
series_example >= 0

0    False
1    False
2     True
4     True
5     True
Name: example, dtype: bool

In [ ]:
series_example[series_example >= -1 &
               series_example <= 1]

ValueError: ignored

In [ ]:
example_series.agg(['min', 'max'])

min    101
max    107
dtype: int64

##### $\blacktriangleright$ Создание объекта `Series`: список

In [ ]:
import pandas as pd
import numpy as np

In [ ]:
xample_series = pd.Series(data = [1]*100, dtype='float', name = 'first' )
xample_series


0     1.0
1     1.0
2     1.0
3     1.0
4     1.0
     ... 
95    1.0
96    1.0
97    1.0
98    1.0
99    1.0
Name: first, Length: 100, dtype: float64

In [ ]:
example_series = pd.Series(data = [1, 4, 7, 6], 
                           dtype = 'int64', 
                           name = 'value')
example_series

0    1
1    4
2    7
3    6
Name: value, dtype: int64

$\color{green}{{Комментарий:}}$ создали объект `Series` с именем `value`. Первый столбец представляет собой *индексные метки* объекта, а второй столбец содержит *значения* объекта. Указывается также тип данных значений объекта `Series`. 

$\color{cornflowerblue}{{Замечание:}}$ обратим внимание, что значение параметра `name` в заголовке созданного объекта `Series` не отображается, а приводится лишь в служебной информации. Далее будет показано, что при добавлении пободного объекта `Series` в датафрейм значение параметра `name` определяет:
- заголовок соответствующего столбца;
- индекс соответствующей строки.

In [ ]:
list_to_series = [1, 4, 7, 6]
example_series = pd.Series(list_to_series, index = ['one', 'two', 'three', 'four'], dtype = 'float64')
example_series

one      1.0
two      4.0
three    7.0
four     6.0
dtype: float64

In [ ]:
example_series = pd.Series([1,4,7,6], ['one', 'two', 'three', 'four'])
example_series

one      1
two      4
three    7
four     6
dtype: int64

$\color{cornflowerblue}{{Замечание:}}$ обратим внимание, что параметры `data` и `index` явно можно не задавать. В этом случае элементы первого списка будут значениями серии, а элементы второго списка -- метками индекса. Если же в конструктор серии передается только один список, то его элементы будут значениями серии, а индекс примет значения $0,1,2,\ldots$.       

И еще несколько интересных приемов создания объекта `Series` из списка.

In [ ]:
example_series = pd.Series(list('12345'), index=list('abcde'))
example_series

a    1
b    2
c    3
d    4
e    5
dtype: object

In [ ]:
example_series = pd.Series([3]*4)
example_series

0    3
1    3
2    3
3    3
dtype: int64

##### $\blacktriangleright$ Создание объекта `Series`: словарь

Напомним, что словарь


Ключи словаря будут являться метками структуры `Series`, а значения словаря -- значениями структуры.

In [ ]:
three_example_series = pd.Series({0 : 'первый элемент серии', 
                                  1 : 'второй элемент серии',
                                  2 : 'третий элемент серии', 
                                  3 : 'последний элемент серии'})
three_example_series

0       первый элемент серии
1       второй элемент серии
2       третий элемент серии
3    последний элемент серии
dtype: object

##### $\blacktriangleright$ Создание объекта `Series`: массив `ndarray`

Объект `Series` можно создать из массива `ndarray`, генерируемого с помощью функций библиотеки `numpy`.   

In [ ]:
example_series = pd.Series(np.arange(2,6))
example_series

0    2
1    3
2    4
3    5
dtype: int64

$\color{green}{{Комментарий:}}$ создается объект `Series` с целочисленными значениями с шагом $1$ из отрезка $[a;[b]-1]$, где $a, b$ -- параметры функции `arange`, $[\cdot]$ -- целая часть числа.

$\color{cornflowerblue}{{Замечание:}}$ обратим внимание, что при создании массива функцией `arange()` правая граница интервала не включается.

In [ ]:
example_series = pd.Series(np.linspace(2, 6, 7))
example_series

0    2.000000
1    2.666667
2    3.333333
3    4.000000
4    4.666667
5    5.333333
6    6.000000
dtype: float64

$\color{green}{{Комментарий:}}$ создается объект `Series` с количеством значений $n$ из отрезка $[a;b]$, где $a, b, n$ -- параметры функции `linspace(a,b,n)`.

In [ ]:
pd.Series(np.random.seed(12))
example_series = pd.Series(np.random.normal(1,3, size=4))
example_series

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: DeprecationWarning: The default dtype for empty Series will be 'object' instead of 'float64' in a future version. Specify a dtype explicitly to silence this warning.
  """Entry point for launching an IPython kernel.


0    2.418957
1   -1.044278
2    1.727318
3   -4.102207
dtype: float64

$\color{green}{{Комментарий:}}$ создается объект `Series`, значения которого подчиняются нормальному закону распределения с параметрами $E=1$ и $V=3$. Количество значений определяется параметром `size`. 

Метод `seed()` используется для инициализации случайных чисел. Этот применяется для хранения случайного метода генерации одних и тех же случайных чисел при многократном выполнении кода на одной или разных машинах.

$\color{cornflowerblue}{{Замечание:}}$ ознакомиться с имеющимися в модуле `random` законами распределений случайных величин можно [здесь](https://docs.python.org/3/library/random.html). 

##### $\blacktriangleright$ Создание объекта `Series`: скалярная величина

In [ ]:
example_series = pd.read_

In [ ]:
example_series = pd.Series(5, index = np.arange(1,101))
example_series

1      5
2      5
3      5
4      5
5      5
      ..
96     5
97     5
98     5
99     5
100    5
Length: 100, dtype: int64

##### $\blacktriangleright$ Работа с элементами Series

$\color{cornflowerblue}{\checkmark}$ Доступ к индексу и значениям объекта `Series`

Доступ к индексу и значениям объекта `Series` можно получить с помощью свойств `.index` и `.values`, соответственно.




In [ ]:
example_series = pd.Series(np.arange(2,6))
example_series

0    2
1    3
2    4
3    5
dtype: int64

In [ ]:
example_series.unique()

array([2, 3, 4, 5])

$\color{green}{{Комментарий:}}$ результатом является объект массива `ndarray` из библиотеки `numpy`.

Узнать тип того или иного объекта можно с помощью метода `type()`. 

In [ ]:
example_series.index

RangeIndex(start=0, stop=4, step=1)

$\color{green}{{Комментарий:}}$ результатом является объект `RangeIndex`, представляющий собой диапазон значений от стартового (**start**) до последнего значения (**stop**) с указанным шагом (**step**). 

`RangeIndex` -- один из типов индексов библиотеки `pandas`.


$\color{cornflowerblue}{\checkmark}$ Размерность объекта `Series`

Количество элементов в объекте `Series` можно определить следующими способами:
- [функцией `len()`](https://www.w3schools.com/python/ref_func_len.asp);
- [свойством `.size`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.size.html);
- [свойством `.shape`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.shape.html).

In [ ]:
len(example_series)

4

In [ ]:
example_series.size

4

In [ ]:
example_series.shape[0]

4

$\color{green}{{Комментарий:}}$ обратим внимание, что в общем случае свойство `shape` возвращает кортеж (тип данных `tuple`), состоящий из двух чисел. Здесь же приводится лишь первое значение этого кортежа, соответствующий количеству элементов в объекте `Series`.

Напомним, что доступ к элементам кортежа осуществляется с помощью квадратных скобок `[]`. Следовательно, при необходимости, чтобы получить размерность объекта `Series` с помощью свойства `shape` надо указать `example_series.shape[0]`

$\color{cornflowerblue}{{Замечание:}}$ в общем случае для определения размерности объектов `Series` или `DataFrame` рекомендуется пользоваться свойством `shape`, поскольку оно сразу возвращает количество строк и столбцов (в случае `DataFrame`).

$\color{cornflowerblue}{\checkmark}$ Вывод значений объекта `Series`

В библиотеке `pandas` имеются случающие методы для вывода значений объекта `Series` или `DataFrame`.

|Метод|Описание|
|--:|:--|
|[`.head()`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.head.html)|возвращает **первые** $n$ строк объекта. По умолчанию $n=5$|
|[`.tail()`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.tail.html)|возвращает **последние** $n$ строк объекта. По умолчанию $n=5$|
|[`.take()`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.take.html)|возвращает строки с указанными позиционными индексами вдоль выбранной оси|

In [ ]:
example_series = pd.Series(data = [1,3,2,6,4,7,8,4,5,6,7,4], 
                           index = np.arange(30,42))
example_series

30    1
31    3
32    2
33    6
34    4
35    7
36    8
37    4
38    5
39    6
40    7
41    4
dtype: int64

In [ ]:
example_series.head(7)

30    1
31    3
32    2
33    6
34    4
35    7
36    8
dtype: int64

In [ ]:
example_series.tail(3)

39    6
40    7
41    4
dtype: int64

In [ ]:
example_series

30    1
31    3
32    2
33    6
34    4
35    7
36    8
37    4
38    5
39    6
40    7
41    4
dtype: int64

In [ ]:
example_series.take(np.arange(1,5))

31    3
32    2
33    6
34    4
dtype: int64

In [ ]:
example_series.sample(3)

38    5
35    7
33    6
dtype: int64

$\color{cornflowerblue}{{Замечание:}}$ обратим внимание, что метод `take()` вернул строки с номерами 1, 2, 3 и 4, что не соответствует индексу рассматриваемого объекта `Series`. Иными словами, данный метод возвращает строки согласно их положению в объекте, а не индексу.

$\color{cornflowerblue}{\checkmark}$ Доступ к данным в объекте `Series`

Для получения значений объекта `Series` пользуются, как правило, следующими двумя способами:
- по **метке индекса**;
- по **позиции с началом отчета в 0**.

$\color{cornflowerblue}{\checkmark\checkmark}$ Доступ к данным в объекте `Series` **по метке индекса**

Доступ **по меткам индекса** предполагает поиск элементов объекта `Series`, метки индекса которых совпадают с заданными. Если задаются отсутствующие метки, то будет сгенерировано исключение `KeyError`. При этом метки индекса могут задаваться в виде:
- отдельных символов `['a']`;
- списка список меток `['a', 'b', 'c']`;

Доступ к элементам объекта по **метке индекса** реализуется посредством использования:
- оператора `[]`;
- свойства `.ix[]` --- начиная с версии `pandas 0.20.0` это свойство заменено;
- свойства `.loc[]`.

Рассмотрим примеры реализации данного способа доступа к данным.

$\color{cornflowerblue}{\checkmark\checkmark\checkmark}$ Доступ к данным в объекте `Series` по **метке индекса**: оператор `[]`

In [ ]:
example_series = pd.Series(data = [101, 102, 103, 104, 105, 106, 107],
                           index = ['a', 'b', 'c', 'd', '1', '2', '3'])
example_series

a    101
b    102
c    103
d    104
1    105
2    106
3    107
dtype: int64

In [ ]:
example_series['a']

101

In [ ]:
example_series[['1','2']]

1    105
2    106
dtype: int64

In [ ]:
example_series[1]

102

In [ ]:
example_series = (pd.Series(data = [101, 102, 103, 104, 105, 106, 107], 
                          index = np.arange(1,8))
                 )
example_series

1    101
2    102
3    103
4    104
5    105
6    106
7    107
dtype: int64

In [ ]:
example_series[1]

101

$\color{cornflowerblue}{{Замечание:}}$ обратим внимание, что в примере 
```python
example_series[1]

>> 102
```

код отработал не так, как мы ожидали --- результат вывода это вторая строка объекта `Series`. Это пример вывода по **позиции**, который будет рассматриваться далее. Следует быть внимательным при обращении к меткам индекса, так, в рассматриваемом примере метки являются объектами типа  `object` и обращаться к ним необходимо соответствующим образом. 

В дополнении к этому примеру заметим, что если бы индекс у рассматриваемого объекта был целочисленным, то обращение к элементам осуществлялось бы по **метке**. Предлагаем читателю самостоятельно задать объект `Series` с произвольным целочисленным индексом и поэкспериментировать. 


$\color{cornflowerblue}{\checkmark\checkmark\checkmark}$ Доступ к данным в объекте `Series` по **метке индекса**: свойство `.loc[]`

In [ ]:
example_series.loc['a']

101

In [ ]:
lst_ind = ['a', 'b']
example_series.loc[lst_ind]

a    101
b    102
dtype: int64

In [ ]:
example_series

a    101
b    102
c    103
d    104
1    105
2    106
3    107
dtype: int64

In [ ]:
list('123')

['1', '2', '3']

In [ ]:
example_series.loc[list('123')]

1    105
2    106
3    107
dtype: int64

In [ ]:
example_series.loc['1']

105

In [ ]:
#целочисленный индекс --- доступ по метке
example_series1 = pd.Series(data = np.arange(101, 105), 
                            index = [0, 2, 1, 4])
example_series1

0    101
2    102
1    103
4    104
dtype: int64

In [ ]:
example_series1.loc[1]

103

In [ ]:
example_series1[1]

103

$\color{cornflowerblue}{{Замечание:}}$ здесь обратим внимание на то, что в примере 
```python
example_series1.loc[1]

>> 103
```
код отработал ожидаемо --- по **метке** индекса.

$\color{cornflowerblue}{\checkmark\checkmark}$ Доступ к данным в объекте `Series` по **позиции**

Доступ к данным по **позиции** основан на использовании целых чисел для доступа к данных. Если задаются отсутствующие позиции, то будет сгенерировано исключение `IndexError`. При этом позиции могут задаваться в виде:
- отдельных целых чисел `[2]`;
- массива целых чисел `[0, 4, 101]`.

Доступ к элементам объекта по **позиции** реализуется посредством использования:
- оператора `[]`;
- свойства `.iloc[]`.

Рассмотрим примеры реализации данного способа доступа к данным.

$\color{cornflowerblue}{\checkmark\checkmark\checkmark}$ Доступ к данным в объекте `Series` по **позиции**: оператор `[]`

In [ ]:
example_series = pd.Series(data = [101, 102, 103, 104, 105, 106, 107],
                           index = ['a', 'b', 'c', 'd', '1', '2', '3'])
example_series

a    101
b    102
c    103
d    104
1    105
2    106
3    107
dtype: int64

In [ ]:
example_series[0]

101

In [ ]:
example_series[[0, 2]]

a    101
c    103
dtype: int64

$\color{cornflowerblue}{\checkmark\checkmark\checkmark}$ Доступ к данным в объекте `Series` по **позиции**: свойство `.iloc[]`

In [ ]:
example_series

a    101
b    102
c    103
d    104
1    105
2    106
3    107
dtype: int64

In [ ]:
example_series.iloc[0]

101

In [ ]:
example_series.iloc[[1, 2, 3]]

b    102
c    103
d    104
dtype: int64

In [ ]:
#целочисленный индекс -- доступ по позиции
example_series1 = pd.Series(data = np.arange(101, 105), 
                            index = [0, 1, 2, 3])
example_series1

0    101
1    102
2    103
3    104
dtype: int64

In [ ]:
example_series1.iloc[[2, 1]]

2    103
1    102
dtype: int64

$\color{green}{{Комментарий:}}$ во избежании путаницы рекомендуем при реализации доступа к элементам объекта `Series` 
по **метке индекса** применять свойство `.loc[]`, а по **позиции** --- свойство `.iloc[]`. 

Дополнительно об особенностях использования свойств `.loc[]` и `.iloc[]` можно почитать в работе [1](), стр.166.

$\color{cornflowerblue}{\checkmark}$ Создание срезов объекта `Series`

Срез --- объект, представляющий набор меток индекса или позиций, используемый для извлечения некоторой части данных.

Создание срезов или слайсинг (slicing) -- это способ извлечения подмножеств данных из различных структур, в частности из объекта `Series`.     

Объект-срез создается с помощью синтаксиса `start:end:step`, компоненты представляют собой первый элемент (`start`), последний элемент (`end`) и приращение, которое называют еще шагом (`step`). Каждый элемент среза является опциональным и может быть опущен. За счет этого обеспечивается больший контроль над последовательностью отбора элементов (отбирать в прямом или обратном порядке) или интервалом отбора элементов (отбирать каждый элемент, каждый второй элемент и т.д.).

Шаг среза (step) -- целое число, по умолчанию равен 1. Может принимать отрицательные значения.

$\color{red}{Для \,\, запонимания}:$ срез -- это набор меток индекса или позиций, который может передаваться оператору `[]` или свойствам `.loc[]` и `.loc[]` для доступа к данным. 

Рассмотрим примеры создания срезов.

$\color{cornflowerblue}{\checkmark\checkmark}$ Создание срезов объекта `Series`: оператор `[]`


In [ ]:
example_series = pd.Series(data = [101, 102, 103, 104, 105, 106, 107],
                           index = ['a', 'b', 'c', 'd', 1, 2, 3])
example_series

a    101
b    102
c    103
d    104
1    105
2    106
3    107
dtype: int64

In [ ]:
example_series['a':'d']

a    101
b    102
c    103
d    104
dtype: int64

In [ ]:
example_series[1:4:1]

b    102
c    103
d    104
dtype: int64

In [ ]:
example_series[1:4]

b    102
c    103
d    104
dtype: int64

In [ ]:
example_series[1:4:2]

b    102
d    104
dtype: int64

In [ ]:
example_series[4:1:-1]

1    105
d    104
c    103
dtype: int64

In [ ]:
example_series[:4]

a    101
b    102
c    103
d    104
dtype: int64

In [ ]:
example_series[6:0:-1]

3    107
2    106
1    105
d    104
c    103
b    102
dtype: int64

In [ ]:
example_series[1:]

b    102
c    103
d    104
1    105
2    106
3    107
dtype: int64

In [ ]:
#доступ к целочисленной части индекса по позиции, правая граница не включается
example_series[4:6]

1    105
2    106
dtype: int64

$\color{cornflowerblue}{{Замечание:}}$ обратим внимание на следующие особенности работы оператора `[]` со срезом:
1. Доступ к элементам объекта осуществляется и по **меткам индекса**, и по **позиции**;
2. Если доступ осуществляется по
  
  2.1 **меткам индекса**, то **правая граница среза (end) ВКЛЮЧАЕТСЯ** в итоговый результат;

  2.2 **позиции**, то **правая граница среза (end) НЕ ВКЛЮЧАЕТСЯ** в итоговый результат.

$\color{cornflowerblue}{\checkmark\checkmark}$ Создание срезов объекта `Series`: свойство `.loc[]`

In [ ]:
example_series = pd.Series(data = [101, 102, 103, 104, 105, 106, 107],
                           index = ['a', 'b', 'c', 'd', 1, 2, 3])
example_series.loc['a':'d']

a    101
b    102
c    103
d    104
dtype: int64

In [ ]:
#Попытка доступа к целочисленной части индекса приводит к ошибке! Почему? 
#example_series.loc[4:6]

In [ ]:
example_series1 = pd.Series(data = [101, 102, 103, 104, 105, 106, 107],
                           index = np.arange(7))
example_series1

0    101
1    102
2    103
3    104
4    105
5    106
6    107
dtype: int64

In [ ]:
example_series1.loc[1:3]

1    102
2    103
3    104
dtype: int64

$\color{cornflowerblue}{{Замечание:}}$ обратим внимание на следующие особенности работы свойства `.loc[]` со срезом:
1. Доступ к элементам объекта осуществляется только по **меткам индекса**. При этом **правая граница среза (end) ВКЛЮЧАЕТСЯ** в итоговый результат.

$\color{cornflowerblue}{\checkmark\checkmark}$ Создание срезов объекта `Series`: свойство `.iloc[]`

In [ ]:
example_series = pd.Series(data = [101, 102, 103, 104, 105, 106, 107],
                           index = ['a', 'b', 'c', 'd', 1, 2, 3])
example_series.iloc[0:3]

a    101
b    102
c    103
dtype: int64

In [ ]:
example_series1 = pd.Series(data = [101, 102, 103, 104, 105, 106, 107],
                           index = np.arange(7))
example_series1

In [ ]:
example_series1.iloc[1:3]

1    102
2    103
dtype: int64

$\color{cornflowerblue}{{Замечание:}}$ обратим внимание на следующие особенности работы свойства `.iloc[]` со срезом:
1. Доступ к элементам объекта осуществляется только по **позиции**. При этом **правая граница среза (end) НЕ ВКЛЮЧАЕТСЯ** в итоговый результат.

$\color{red}{Для \,\, запонимания}:$ еще раз обратим внимание на тот факт, что оператор `[]` осуществляет доступ к элементам объекта `Series` и по **метке индекса**, и по **позиции**. В случае, если у объекта целочисленный индекс, то использование оператора `[]` может привести к определенной путанице. 

В связи с этим повторим следующую рекомендацию:
- для организации доступа по **меткам индекса** использовать свойство `.loc[]`;
- для организации доступа по **позиции** использовать свойство `.iloc[]`.                      

$\color{OliveDrab}{\widehat{⨆}}$ Упражнение для закрепления материала
```python
example_series1[1:3]
```
Укажите результат?


$\color{cornflowerblue}{\checkmark}$ Фильтрация данных объекта `Series`

$\widehat{DEF.}$ Фильтрация данных -- вывод нужных пользователю данных, в результате созданного им запроса.

Выше рассматривался вопрос о доступе к элементам объекта `Series` с помощью индекса или позиции. В этой части материала остановимся на вопросе отбора (фильтрации) элементов объекта на основе их значений.

Для решения этого вопроса библиотека `pandas` предлагает воспользоваться логическим отбором (отбором на основе булевых значений).  Логический отбор применяет логическое выражение к значениям объекта `Series` и возвращает **новую серию той же размерности и с метками индекса**, как у исходного объекта. Значения возвращаемого нового объекта представляют собой результат применения заданного логического выражения к каждому значению исходного объекта и равны

$
\left[
\begin{array}{ll}
True, \text{если значение объекта соответствует заданному логическому выражению,}\\
False, \text{если значение объекта не соответствует заданному логическому выражению.}
\end{array}  
\right.  
$

Далее этот результат (новый объект `Series`, состоящий из логических значений) можно использовать для извлечения лишь тех значений исходного объекта, для которых получено значение `True`.


Логическое выражение может задаваться любые доступными способами, например, непосредственно или с помощью $\lambda$-функции.

Если задаваемое логическое выражение состоит из нескольких условий, то каждое из них нужно взять в круглые скобки `()` и соединить необходимыми двоичными операторами. Например, $(\text{условие 1}) \, \&  \, (\text{условие 2})$. 

|Двоичный оператор|Комментарий|
|--:|:--|
|`~`|двоичная инверсия (логическое отрицание)|
|`&`|двоичное `И`|
|$|$|двоичное `ИЛИ`|
|`^`|двоичное исключающее `ИЛИ`|

$\color{cornflowerblue}{{Замечание:}}$ следует запомнить, что ключевые слова `and` и `or` с булевыми массивами не работают. Используйте вместо них `&` (и) и `|` (или). Дополнительно о данной особенности можно почитать [здесь](https://overcoder.net/q/63576/%D0%BB%D0%BE%D0%B3%D0%B8%D1%87%D0%B5%D1%81%D0%BA%D0%B8%D0%B5-%D0%BE%D0%BF%D0%B5%D1%80%D0%B0%D1%82%D0%BE%D1%80%D1%8B-%D0%B4%D0%BB%D1%8F-%D0%BB%D0%BE%D0%B3%D0%B8%D1%87%D0%B5%D1%81%D0%BA%D0%BE%D0%B3%D0%BE-%D0%B8%D0%BD%D0%B4%D0%B5%D0%BA%D1%81%D0%B8%D1%80%D0%BE%D0%B2%D0%B0%D0%BD%D0%B8%D1%8F-%D0%B2-pandas) или в [2, стр.85].

Реализуется логический отбор либо с помощью оператора `[]`, либо с помощью свойства `.loc[]`. 

Рекомендуем также обратить внимание на следующие полезные методы

|Метод|Комментарий|
|--:|:--|
|[`.all()`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.all.html)|возвращает `True`, если **все элементы** объекта удовлетворяют заданному условию, иначе `False`|
|[`.any()`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.any.html)|возвращает `True`, если **хотя бы один элемент** объекта удовлетворяют заданному условию, иначе `False`|
|[`.sum()`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.sum.html)|возвращает сумму значений|
|[`.isin()`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.isin.html)|возвращает булев массив, показывающий, содержится ли каждое принадлежащее `Series` значение в переданной последовательности|
|[`.unique()`](https://pandas.pydata.org/docs/reference/api/pandas.Series.unique.html)|возвращает массив уникальных значений в Series и возвращает их в порядке появления|
|[`.nunique()`](https://pandas.pydata.org/docs/reference/api/pandas.Series.nunique.html)|возвращает количество уникальных элементов в объекте|
|[`.value_counts()`](https://pandas.pydata.org/docs/reference/api/pandas.Series.value_counts.html)|возвращает объект Series, который содержит уникальное значение в качестве индекса и его частоту в качестве соответствующего значения|
|[`.nlargest()`](https://pandas.pydata.org/docs/reference/api/pandas.Series.nlargest.html)|возвращает заданное $n$ наибольших значений объекта. По умолчанию $n=5$|
|[`.nsmalest()`](https://pandas.pydata.org/docs/reference/api/pandas.Series.nsmallest.html)|возвращает заданное $n$ наименьших значений объекта. По умолчанию $n=5$|

Использование данных методов может быть полезным, в частности, при оценке корретности имеющихся данных, например, положительности данных о росте и т.п.  

Рассмотрим примеры фильтрации значений объекта `Series` с помощью логического отбора.  



In [ ]:
example_series = pd.Series(data = [101, 102, 103, 104, 105, 106, 107],
                           index = ['a', 'b', 'c', 'd', 1, 2, 3])
example_series

a    101
b    102
c    103
d    104
1    105
2    106
3    107
dtype: int64

In [ ]:
example_series > 103

a    False
b    False
c    False
d     True
1     True
2     True
3     True
dtype: bool

In [ ]:
example_series[example_series > 103]

d    104
1    105
2    106
3    107
dtype: int64

In [ ]:
(example_series > 103) & (example_series <= 106)

a    False
b    False
c    False
d     True
1     True
2     True
3    False
dtype: bool

In [ ]:
example_series[(example_series > 103) & 
               (example_series <= 106)]

d    104
1    105
2    106
dtype: int64

In [ ]:
example_series[~((example_series > 103) & 
               (example_series <= 106))]

a    101
b    102
c    103
3    107
dtype: int64

In [ ]:
~((example_series > 103) & (example_series <= 106))

a     True
b     True
c     True
d    False
1    False
2    False
3     True
dtype: bool

In [ ]:
example_series.loc[~((example_series > 103) & 
               (example_series <= 106))]

a    101
b    102
c    103
3    107
dtype: int64

In [ ]:
example_series[lambda x: ((x>103) & 
                         (x<106))]

d    104
1    105
dtype: int64

In [ ]:
example_series1 = pd.Series(data = [173, 0, -183, 190, 184, 183, 182],
                           index = np.arange(7))
example_series1


0    173
1      0
2   -183
3    190
4    184
5    183
6    182
dtype: int64

In [ ]:
(example_series1 > 0).all()

False

In [ ]:
(example_series1 >= 190).any().sum()

1

In [ ]:
example_series1 = pd.Series(data = ['м', 'ж', 'м', 'ж', 'мж'],
                           index = np.arange(5))
example_series1


0     м
1     ж
2     м
3     ж
4    мж
dtype: object

In [ ]:
example_series1.isin(['м', 'ж'])

0     True
1     True
2     True
3     True
4    False
dtype: bool

In [ ]:
example_series1.loc[example_series1.isin(['м', 'ж'])]

Series([], dtype: object)

In [ ]:
example_series1.isin(['м', 'ж']))

1

In [ ]:
example_series1.unique()

array(['м', 'ж', 'мж'], dtype=object)

In [ ]:
example_series1.value_counts()

м     2
ж     2
мж    1
dtype: int64

In [ ]:
example_series1.nunique()

3

##### $\S$ Выравнивание данных

Выравнивание осуществляет автоматическое согласование связанных друг с другом значений в нескольких объектах `Series` на основе меток индекса. Наглядно процедура выравнивания демострируется при проведении арифметических операций над объектами `Series` с разными метками индексами. Так, например, если при сложении двух объектов
обнаруживаются различные пары индексов, то результирующий индекс будет
объединением индексов.    

Метки в индексах не обязательны для выравнивания. Если выравнивания не происходит, то `pandas` в качестве результата возвращает значение `NaN`.

Метки в индексе не обязательно должны быть уникальными. Операция выравнивания фактически образует декартово произведение меток в двух объектах `Series`. Если в первой серии имеется $n$ одинаковых меток, а во второй серии -- $m$ одинаковых меток, то в итоговом результате будет $n\cdot m$ одинаковых меток. 


Рассмотрим примеры.





In [ ]:
ser1 = pd.Series(data = np.arange(1,5),
                 index = np.arange(1,5))
ser1

1    1
2    2
3    3
4    4
dtype: int64

In [ ]:
ser2 = pd.Series(data = np.arange(1,5),
                 index = np.arange(4,0, -1))
ser2

4    1
3    2
2    3
1    4
dtype: int64

In [ ]:
ser1+ser2

1    5
2    5
3    5
4    5
dtype: int64

In [ ]:
ser3 = pd.Series(data = np.arange(1,6),
                 index = np.arange(1,6))
ser3

1    1
2    2
3    3
4    4
5    5
dtype: int64

In [ ]:
ser1+ser3

1    2.0
2    4.0
3    6.0
4    8.0
5    NaN
dtype: float64

In [ ]:
ser4 = pd.Series([1.0, 2.0, 3.0], index=[1, 1, 2])
ser4

1    1.0
1    2.0
2    3.0
dtype: float64

In [ ]:
ser5 = pd.Series([1.0, 2.0, 3.0], index=[1, 1, 1])
ser5

1    1.0
1    2.0
1    3.0
dtype: float64

In [ ]:
ser4+ser5

1    2.0
1    3.0
1    4.0
1    3.0
1    4.0
1    5.0
2    NaN
dtype: float64

#### Маркированная многомерная структура данных  `DataFrame`

`DataFrame` --- это проиндексированный многомерный массив значений, в котором каждый столбец является структурой `Series`. Каждая строка датафрейма описывает различные характеристики исследуемого субъекта и каждый столбец может представлять различные типы данных.

Рассмотрим следующие вопросы по работе с датафреймом:
- создание датафрейма;
- определение размерности датафрейма;
- назначение и изменение столбцов в датафрейме;
- отбор строк и столбцов датафрейма;
- создание среза датафрейма.

#### Создание датафрейма

Конструктор класса `DataFrame` выглядит так:

```python
DataFrame(data=None, index=None, columns=None, dtype=None, copy=False)
```

Параметры конструктора:

|Параметр|Комментарий|
|--:|:--|
|`data`: `ndarray`, `dict` или `DataFrame`; значение по умолчанию: `None`|Данные, на базе которых будет создан `DataFrame`|
|`index`: одномерный массив; значение по умолчанию: `None`|Список меток для записей (имена строк таблицы)|
|`columns`: одномерный массив; значение по умолчанию: `None`|Список меток для полей (имена столбцов таблицы)|
|`dtype`: `numpy.dtype`; значение по умолчанию: `None`|Объект, определяющий тип данных|
|`copy`: `bool`; значение по умолчанию: `False`|Если параметр равен `True`, то будет создана копия массива данных|

Структуру DataFrame можно создать на базе следующих типов данных:
- словарь (dict), в качестве элементов которого могут выступать: одномерные ndarray, списки, другие словари, структуры Series;
- двумерный ndarray;
- структура Series;
- другой DataFrame;
- из внешних источников (работа с внешними источниками).

##### Создание датафрейма: серия

In [ ]:
series1_to_df = pd.Series([1,2,3,4], name = 'ser1')
series2_to_df = pd.Series([5,6,7,8], name = 'ser2')

In [ ]:
#из одного объекта Series
df = pd.DataFrame(series1_to_df)
df

ser1
0     1
1     2
2     3
3     4

In [ ]:
#из нескольких объектов Series
df = pd.DataFrame([series1_to_df, series2_to_df])
df 

0  1  2  3
ser1  1  2  3  4
ser2  5  6  7  8

$\color{cornflowerblue}{{Замечание:}}$ Для того, чтобы транспонировать датафрейм можно воспользоваться свойством [.T](https://pandas.pydata.org/pandas-docs/version/0.25.0/reference/api/pandas.DataFrame.T.html).

In [ ]:
df_T = df.T
df_T

ser1  ser2
0     1     5
1     2     6
2     3     7
3     4     8

##### Создание датафрейма: словарь

In [ ]:
dict_to_df = {'col1':[1,'2',3,4],
              'col2':[5,6,7,8]
              }
df = pd.DataFrame(dict_to_df, 
                  index = np.arange(1,5))
df


col1  col2
1    1     5
2    2     6
3    3     7
4    4     8

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4 entries, 1 to 4
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   col1    4 non-null      object
 1   col2    4 non-null      int64 
dtypes: int64(1), object(1)
memory usage: 96.0+ bytes


In [ ]:
dict_to_df = {'col1': pd.Series(data = np.arange(1,5)),
              'col2': pd.Series(data = np.arange(1,6))
              }

In [ ]:
#выравнивание датафрейма
df = pd.DataFrame(dict_to_df)
df

col1  col2
0   1.0     1
1   2.0     2
2   3.0     3
3   4.0     4
4   NaN     5

In [ ]:
dict_to_df = [{'product': 'Апельсин', 'price': 2, 'count': 3},
             {'product': 'Мандарин', 'price': 4, 'count': 5, 'count1': 5}]

In [ ]:
df = pd.DataFrame(dict_to_df)
df

product  price  count  count1
0  Апельсин      2      3     NaN
1  Мандарин      4      5     5.0

##### Создание датафрейма: результатов функция `numpy`

In [ ]:
data = np.array([[10, 11], [20, 21]]) 
df = pd.DataFrame(data)
df

0   1
0  10  11
1  20  21

##### Создание датафрейма: из других датафреймов

In [ ]:
purchase_1 = pd.Series({'Name': 'JJ',
                        'Item': 'A',
                        'Cost': 22.00})
purchase_2 = pd.Series({'Name': 'KK',
                        'Item': 'B',
                        'Cost': 22.50})

dfn = pd.DataFrame([purchase_1, purchase_2], index=['Store X', 'Store Y'])

dfn = dfn.append(pd.Series(data={'Cost': 30.00, 'Item': 'C','Name': 'TT'}, name='00Store Y'))
dfn


Name Item  Cost
Store X     JJ    A  22.0
Store Y     KK    B  22.5
00Store Y   TT    C  30.0

##### Создание датафрейма: игрушечные данные, сбор промежуточных данных

In [ ]:
pd.DataFrame(np.random.rand(4, 8), columns=list('abcdefgh'))

a         b         c         d         e         f         g  \
0  0.016400  0.533106  0.556849  0.880720  0.377556  0.105608  0.645108   
1  0.405220  0.039243  0.624034  0.787297  0.675635  0.456310  0.552630   
2  0.843469  0.137704  0.409594  0.285665  0.456518  0.841911  0.410225   
3  0.008055  0.100904  0.803350  0.725829  0.782117  0.302863  0.953277   

          h  
0  0.647871  
1  0.119901  
2  0.807062  
3  0.311958

##### Работа с внешними источниками данных



###### Работа с данными в формате [`CSV`](https://ru.wikipedia.org/wiki/CSV#:~:text=CSV%20(%D0%BE%D1%82%20%D0%B0%D0%BD%D0%B3%D0%BB.,%D0%A4%D0%BE%D1%80%D0%BC%D0%B0%D1%82%20CSV%20%D1%81%D1%82%D0%B0%D0%BD%D0%B4%D0%B0%D1%80%D1%82%D0%B8%D0%B7%D0%B8%D1%80%D0%BE%D0%B2%D0%B0%D0%BD%20%D0%BD%D0%B5%20%D0%BF%D0%BE%D0%BB%D0%BD%D0%BE%D1%81%D1%82%D1%8C%D1%8E.)

Предположим, что исследуемые данные хранятся в `csv`-файле. Требуется загрузить данные из этого файла в соответствующую структуру библиотеки `pandas` для последующей обработки. 

Для загрузки данных из `csv`-файла воспользуемся методом `read_csv()`. В связи с тем, что в исходных данных может быть множество различных особенностей этот метод имеет множество аргументов загрузки. Перечислим наиболее актуальные из них:
- **`filepath_or_buffer`** --- строка, содержащая путь до файла или буфера c данными в формате `CSV`;
- **`sep`**, **`delimiter`** --- строка, задающая разделитель между полями загружаемых  данных. Она может содержать последовательность символов или регулярное выражение. Распространенными разделителями также являются `;` и `\t`. По умолчанию аргумент `sep=','`, а `delimiter=None`. При задании эти два аргумента являются эквивалентными, однако, на практике чаще используется аргумент **`sep`**;
- **`usecols`** --- список загружаемых столбцов;
- **`header`** --- номер строки, содержащей имена столбцов. По умолчанию равен **`header = 0`** (первая строка). Если строки-заголовка нет, то аргумент должен быть равен **`header = None`**. При этом имена столбцов можно задать в параметре  **`names`**. Обратим внимание, что если вначале файла идут пустые строки или строки с комментариями, а далее строка с именами столбцов, то необходимо положить **`skip_blank_lines=True`**;
- **`names`** --- список имен столбцов таблицы, используется, если в файле нет
строки с именами столбцов и аргумент **`header=None`**. По умолчанию **`names=None`**;
- **`index_col`** --- номера или имена столбцов, трактуемых как индекс строк в результирующем объекте. Может быть задан один номер (имя) или список номеров (имен), определяющий иерархический индекс;
- **`usecols`** --- список имен или индексов столбцов, которые подлежат загрузке;
- **`dtype`** --- тип загружаемых данных в целом или словарь, в котором ключами будут имена столбцов, а значениями --- типы столбцов;
- **`nrows`** --- количество читаемых строк от начала файла;
- **`iterator`** --- возвращает объект TextParser для чтения файла порциями;
- **`chunksize`** --- размер порции при итерировании.

##### Работа с форматом `csv`

In [ ]:
df = pd.read_csv('data/real_estate_data.csv',
                 delimiter = '\t'
                 #usecols = ['total_images', 'last_price']
                 #nrows = 3 
                 )
df

total_images  last_price  total_area first_day_exposition  rooms  \
0                20  13000000.0      108.00  2019-03-07T00:00:00      3   
1                 7   3350000.0       40.40  2018-12-04T00:00:00      1   
2                10   5196000.0       56.00  2015-08-20T00:00:00      2   
3                 0  64900000.0      159.00  2015-07-24T00:00:00      3   
4                 2  10000000.0      100.00  2018-06-19T00:00:00      2   
...             ...         ...         ...                  ...    ...   
23694             9   9700000.0      133.81  2017-03-21T00:00:00      3   
23695            14   3100000.0       59.00  2018-01-15T00:00:00      3   
23696            18   2500000.0       56.70  2018-02-11T00:00:00      2   
23697            13  11475000.0       76.75  2017-03-28T00:00:00      2   
23698             4   1350000.0       32.30  2017-07-21T00:00:00      1   

       ceiling_height  floors_total  living_area  floor is_apartment  ...  \
0                2.70          16.0         51.0      8          NaN  ...   
1                 NaN          11.0         18.6      1          NaN  ...   
2                 NaN           5.0         34.3      4          NaN  ...   
3                 NaN          14.0          NaN      9          NaN  ...   
4                3.03          14.0         32.0     13          NaN  ...   
...               ...           ...          ...    ...          ...  ...   
23694            3.70           5.0         73.3      3          NaN  ...   
23695             NaN           5.0         38.0      4          NaN  ...   
23696             NaN           3.0         29.7      1          NaN  ...   
23697            3.00          17.0          NaN     12          NaN  ...   
23698            2.50           5.0         12.3      1          NaN  ...   

       kitchen_area  balcony        locality_name  airports_nearest  \
0             25.00      NaN      Санкт-Петербург           18863.0   
1             11.00      2.0       посёлок Шушары           12817.0   
2              8.30      0.0      Санкт-Петербург           21741.0   
3               NaN      0.0      Санкт-Петербург           28098.0   
4             41.00      NaN      Санкт-Петербург           31856.0   
...             ...      ...                  ...               ...   
23694         13.83      NaN      Санкт-Петербург           24665.0   
23695          8.50      NaN                Тосно               NaN   
23696           NaN      NaN     село Рождествено               NaN   
23697         23.30      2.0      Санкт-Петербург           39140.0   
23698          9.00      NaN  поселок Новый Учхоз               NaN   

      cityCenters_nearest  parks_around3000  parks_nearest  ponds_around3000  \
0                 16028.0               1.0          482.0               2.0   
1                 18603.0               0.0            NaN               0.0   
2                 13933.0               1.0           90.0               2.0   
3                  6800.0               2.0           84.0               3.0   
4                  8098.0               2.0          112.0               1.0   
...                   ...               ...            ...               ...   
23694              4232.0               1.0          796.0               3.0   
23695                 NaN               NaN            NaN               NaN   
23696                 NaN               NaN            NaN               NaN   
23697             10364.0               2.0          173.0               3.0   
23698                 NaN               NaN            NaN               NaN   

       ponds_nearest  days_exposition  
0              755.0              NaN  
1                NaN             81.0  
2              574.0            558.0  
3              234.0            424.0  
4               48.0            121.0  
...              ...              ...  
23694          381.0              NaN  
23695            NaN             45.0  
23696  

##### Работа с форматом `xlsx`  

In [ ]:
df_from_xlsx = pd.read_excel('data/lst_abitur.xlsx', sheet_name = 'Лист1' 
                             )
df_from_xlsx

a  b  c
0  1  2  4
1  5  6  7

##### Загрузка данных: нескольких файлов 

Создание датафрейма из нескольких файлов (по строкам)

Допустим, ваш набор данных распределен по нескольким файлам, но вы хотите собрать набор данных в один датафрейм. Будем также считать, что вы знаете заголовки требуемых столбцов:

|Признак|Комментарий|
|--:|:--|
|`user_id`|уникальный идентификатор клиента|
|`surname`|фамилия клиента|
|`name`|имя клиента|
|`date_of_birth`|дата рождения клиента|

Предположим также, что исходные данные хранятся в файлах `user1.csv`, `user2.csv` и `user3.csv`.

Вы можете прочитать каждый CSV-файл в свой собственный DataFrame, объединить их вместе, а затем удалить исходные DataFrames, но это будет неэффективно с точки зрения использования памяти и потребует большого количества кода.

Лучшим решением является использование встроенного модуля `glob`.

Однако прежде прочтем один файл и посмотрим на его структуру.

In [ ]:
read_columns = ['user_id', 'surname', 'name', 'date_of_birth']

In [ ]:
pd.read_csv('user1.csv', 
            usecols = read_columns)

user_id    surname     name date_of_birth
0       11    Сидоров     Иван    1982-05-04
1       12     Иванов  Николай    1983-09-12
2       13     Петров     Петр    1983-12-12
3       14  Гришечкин  Алексей    1985-01-13
4       15  Афанасьев   Сергей    1986-08-15
5       16      Мимин   Андрей    1985-01-15

In [ ]:
from glob import glob

Вы можете передать шаблон в `glob()`, включая подстановочные знаки, и он вернет список всех файлов, соответствующих этому шаблону.

В этом случае метод `glob` в папке `data` ищет все `CSV`-файлы, начинающиеся со слова `user`.

In [ ]:
stock_files = sorted(glob('user*.csv'))
#stock_files = ['1.csv', 'test.csv']
stock_files

['user1.csv', 'user2.csv', 'user3.csv']

`glob` возвращает имена файлов в произвольном порядке, поэтому мы отсортировали список с помощью встроенной в Python функции `sorted()`.

Затем мы можем использовать выражение генератора для чтения каждого из файлов с помощью `read_csv()` и передать результаты функции `concat()`, которая объединит строки в один `DataFrame`

In [ ]:
pd.concat((pd.read_csv(file) for file in stock_files))

Unnamed: 0  user_id    surname       name date_of_birth     city
0           0       11    Сидоров       Иван    1982-05-04      NaN
1           1       12     Иванов    Николай    1983-09-12      NaN
2           2       13     Петров       Петр    1983-12-12      NaN
3           3       14  Гришечкин    Алексей    1985-01-13      NaN
4           4       15  Афанасьев     Сергей    1986-08-15      NaN
5           5       16      Мимин     Андрей    1985-01-15      NaN
0           0       23    Петрова  Анастасия    1986-02-12      NaN
1           1       24   Сидорова    Людмила    1984-12-25      NaN
2           2       25    Иванова      Ирина    1986-04-15      NaN
3           3       26    Фирсова       Анна    1985-06-23      NaN
0           0       33     Малкин   Афанасий    1986-05-12     Орел
1           1       34   Протасов      Тарас    1992-12-25  Воронеж
2           2       35   Кочергин   Григорий    1987-02-25     Елец

К сожалению, теперь в индексе есть повторяющиеся значения. Чтобы избежать этого, мы можем указать функции `concat()` игнорировать индекс и вместо этого использовать целочисленный индекс по умолчанию.

In [ ]:
pd.concat((pd.read_csv(file, usecols = read_columns) for file in stock_files), ignore_index=True)

user_id    surname       name date_of_birth
0        11    Сидоров       Иван    1982-05-04
1        12     Иванов    Николай    1983-09-12
2        13     Петров       Петр    1983-12-12
3        14  Гришечкин    Алексей    1985-01-13
4        15  Афанасьев     Сергей    1986-08-15
5        16      Мимин     Андрей    1985-01-15
6        23    Петрова  Анастасия    1986-02-12
7        24   Сидорова    Людмила    1984-12-25
8        25    Иванова      Ирина    1986-04-15
9        26    Фирсова       Анна    1985-06-23
10       33     Малкин   Афанасий    1986-05-12
11       34   Протасов      Тарас    1992-12-25
12       35   Кочергин   Григорий    1987-02-25

Создание датафрейма из нескольких файлов (по столбцам)

Предположим теперь, что каждый файл будет содержать столбцы из единого набора данных. Давайте соберем единый датафрейм с столбцов, содержащихся в разных файлах.

In [ ]:
pd.read_csv('user1.csv').head()

Unnamed: 0  user_id    surname     name date_of_birth
0           0       11    Сидоров     Иван    1982-05-04
1           1       12     Иванов  Николай    1983-09-12
2           2       13     Петров     Петр    1983-12-12
3           3       14  Гришечкин  Алексей    1985-01-13
4           4       15  Афанасьев   Сергей    1986-08-15

In [ ]:
pd.read_csv('session_user1.csv').head()

Unnamed: 0  session_id           time_start             time_end
0           0        1111  1982-05-04:12-35-56  1982-05-04:13-05-34
1           1        1222  1983-09-12:23-01-06  1983-09-12:23-05-56
2           2        1333  1983-12-12:05-25-13  1983-12-12:06-00-06
3           3        1444  1985-01-13:10-39-24  1985-01-13:10-40-56
4           4        1555  1986-08-15:12-37-42  1986-08-15:14-18-19

Создадим верную последовательность файлов для чтения. 

In [ ]:
lst_file = ['user1.csv', 'session_user1.csv']

С помощью метода `concat()` выполним конкатенацию данных вдоль оси столбцов

In [ ]:
df = pd.concat((pd.read_csv(file) for file in lst_file), axis='columns').head()
df

Unnamed: 0  user_id    surname     name date_of_birth  Unnamed: 0  \
0           0       11    Сидоров     Иван    1982-05-04           0   
1           1       12     Иванов  Николай    1983-09-12           1   
2           2       13     Петров     Петр    1983-12-12           2   
3           3       14  Гришечкин  Алексей    1985-01-13           3   
4           4       15  Афанасьев   Сергей    1986-08-15           4   

   session_id           time_start             time_end  
0        1111  1982-05-04:12-35-56  1982-05-04:13-05-34  
1        1222  1983-09-12:23-01-06  1983-09-12:23-05-56  
2        1333  1983-12-12:05-25-13  1983-12-12:06-00-06  
3        1444  1985-01-13:10-39-24  1985-01-13:10-40-56  
4        1555  1986-08-15:12-37-42  1986-08-15:14-18-19

In [ ]:
df.drop(columns = ['Unnamed: 0'], inplace =True)

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 7 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   user_id        5 non-null      int64 
 1   surname        5 non-null      object
 2   name           5 non-null      object
 3   date_of_birth  5 non-null      object
 4   session_id     5 non-null      int64 
 5   time_start     5 non-null      object
 6   time_end       5 non-null      object
dtypes: int64(2), object(5)
memory usage: 408.0+ bytes


Создание датафреймов из нескольких файлов 

In [ ]:
list_dict = {'Таблица "Информация о посещениях сайта"' : '/data/visits_log.csv', 
            'Таблица "Информация о заказах"' : '/data/orders_log.csv',
             'Таблица "Информация о расходах на маркетинг"' : '/data/costs.csv',
            }
df_list=[]
for i in list_dict.values():
    df_list.append(pd.read_csv(i))
    print('Прочитан файл {}'.format(i))
globals()['df'+'_clients'] = df_list[0]

#### Просмотр фрагмента данных

Методы `head()`, `tail()`, `sample()` и `take()`.

In [ ]:
df.head(2)

total_images  last_price  total_area first_day_exposition  rooms  \
0            20  13000000.0       108.0  2019-03-07T00:00:00      3   
1             7   3350000.0        40.4  2018-12-04T00:00:00      1   

   ceiling_height  floors_total  living_area  floor is_apartment  ...  \
0             2.7          16.0         51.0      8          NaN  ...   
1             NaN          11.0         18.6      1          NaN  ...   

   kitchen_area  balcony    locality_name  airports_nearest  \
0          25.0      NaN  Санкт-Петербург           18863.0   
1          11.0      2.0   посёлок Шушары           12817.0   

  cityCenters_nearest  parks_around3000  parks_nearest  ponds_around3000  \
0             16028.0               1.0          482.0               2.0   
1             18603.0               0.0            NaN               0.0   

   ponds_nearest  days_exposition  
0          755.0              NaN  
1            NaN             81.0  

[2 rows x 22 columns]

In [ ]:
df.tail(2)

total_images  last_price  total_area first_day_exposition  rooms  \
23697            13  11475000.0       76.75  2017-03-28T00:00:00      2   
23698             4   1350000.0       32.30  2017-07-21T00:00:00      1   

       ceiling_height  floors_total  living_area  floor is_apartment  ...  \
23697             3.0          17.0          NaN     12          NaN  ...   
23698             2.5           5.0         12.3      1          NaN  ...   

       kitchen_area  balcony        locality_name  airports_nearest  \
23697          23.3      2.0      Санкт-Петербург           39140.0   
23698           9.0      NaN  поселок Новый Учхоз               NaN   

      cityCenters_nearest  parks_around3000  parks_nearest  ponds_around3000  \
23697             10364.0               2.0          173.0               3.0   
23698                 NaN               NaN            NaN               NaN   

       ponds_nearest  days_exposition  
23697          196.0            602.0  
23698            NaN              NaN  

[2 rows x 22 columns]

In [ ]:
df.sample(2)

total_images  last_price  total_area first_day_exposition  rooms  \
20416             7  14780000.0        77.1  2018-10-25T00:00:00      2   
22847             3   2100000.0        32.2  2017-06-02T00:00:00      1   

       ceiling_height  floors_total  living_area  floor is_apartment  ...  \
20416             NaN           4.0         33.9      4          NaN  ...   
22847             NaN           5.0         18.4      3          NaN  ...   

       kitchen_area  balcony      locality_name  airports_nearest  \
20416          15.8      NaN    Санкт-Петербург           18617.0   
22847           5.4      0.0  посёлок Понтонный           31458.0   

      cityCenters_nearest  parks_around3000  parks_nearest  ponds_around3000  \
20416             17465.0               1.0         1767.0               3.0   
22847             33170.0               0.0            NaN               0.0   

       ponds_nearest  days_exposition  
20416          111.0            170.0  
22847            NaN             88.0  

[2 rows x 22 columns]

In [ ]:
df.take([1,2,3])

total_images  last_price  total_area first_day_exposition  rooms  \
1             7   3350000.0        40.4  2018-12-04T00:00:00      1   
2            10   5196000.0        56.0  2015-08-20T00:00:00      2   
3             0  64900000.0       159.0  2015-07-24T00:00:00      3   

   ceiling_height  floors_total  living_area  floor is_apartment  ...  \
1             NaN          11.0         18.6      1          NaN  ...   
2             NaN           5.0         34.3      4          NaN  ...   
3             NaN          14.0          NaN      9          NaN  ...   

   kitchen_area  balcony    locality_name  airports_nearest  \
1          11.0      2.0   посёлок Шушары           12817.0   
2           8.3      0.0  Санкт-Петербург           21741.0   
3           NaN      0.0  Санкт-Петербург           28098.0   

  cityCenters_nearest  parks_around3000  parks_nearest  ponds_around3000  \
1             18603.0               0.0            NaN               0.0   
2             13933.0               1.0           90.0               2.0   
3              6800.0               2.0           84.0               3.0   

   ponds_nearest  days_exposition  
1            NaN             81.0  
2          574.0            558.0  
3          234.0            424.0  

[3 rows x 22 columns]

#### Определение размерности датафрейма

Команды `len`, `shape`

In [ ]:
len(df)

23699

In [ ]:
df.shape

(23699, 22)

In [ ]:
type(df.columns)

pandas.core.indexes.base.Index

In [ ]:
len(df.columns)

22

#### Типы данных

In [ ]:
df.dtypes

total_images              int64
last_price              float64
total_area              float64
first_day_exposition     object
rooms                     int64
ceiling_height          float64
floors_total            float64
living_area             float64
floor                     int64
is_apartment             object
studio                     bool
open_plan                  bool
kitchen_area            float64
balcony                 float64
locality_name            object
airports_nearest        float64
cityCenters_nearest     float64
parks_around3000        float64
parks_nearest           float64
ponds_around3000        float64
ponds_nearest           float64
days_exposition         float64
dtype: object

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23699 entries, 0 to 23698
Data columns (total 22 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   total_images          23699 non-null  Int64  
 1   last_price            23699 non-null  Int64  
 2   total_area            23699 non-null  Float64
 3   first_day_exposition  23699 non-null  string 
 4   rooms                 23699 non-null  Int64  
 5   ceiling_height        14504 non-null  Float64
 6   floors_total          23613 non-null  Int64  
 7   living_area           21796 non-null  Float64
 8   floor                 23699 non-null  Int64  
 9   is_apartment          2775 non-null   boolean
 10  studio                23699 non-null  boolean
 11  open_plan             23699 non-null  boolean
 12  kitchen_area          21421 non-null  Float64
 13  balcony               12180 non-null  Int64  
 14  locality_name         23650 non-null  string 
 15  airports_nearest   

In [ ]:
df = df.convert_dtypes()
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23699 entries, 0 to 23698
Data columns (total 22 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   total_images          23699 non-null  Int64  
 1   last_price            23699 non-null  Int64  
 2   total_area            23699 non-null  Float64
 3   first_day_exposition  23699 non-null  string 
 4   rooms                 23699 non-null  Int64  
 5   ceiling_height        14504 non-null  Float64
 6   floors_total          23613 non-null  Int64  
 7   living_area           21796 non-null  Float64
 8   floor                 23699 non-null  Int64  
 9   is_apartment          2775 non-null   boolean
 10  studio                23699 non-null  boolean
 11  open_plan             23699 non-null  boolean
 12  kitchen_area          21421 non-null  Float64
 13  balcony               12180 non-null  Int64  
 14  locality_name         23650 non-null  string 
 15  airports_nearest   

##### Описательные статистики

In [ ]:
df.describe(percentiles=[0.03, 0.25, 0.75, 0.95])

total_images    last_price    total_area         rooms  \
count   23699.000000  2.369900e+04  23699.000000  23699.000000   
unique           NaN           NaN           NaN           NaN   
top              NaN           NaN           NaN           NaN   
freq             NaN           NaN           NaN           NaN   
mean        9.858475  6.541549e+06     60.348651      2.070636   
std         5.682529  1.088701e+07     35.654083      1.078405   
min         0.000000  1.219000e+04     12.000000      0.000000   
3%          0.000000  1.550000e+06     30.000000      1.000000   
25%         6.000000  3.400000e+06     40.000000      1.000000   
50%         9.000000  4.650000e+06     52.000000      2.000000   
75%        14.000000  6.800000e+06     69.900000      3.000000   
95%        20.000000  1.530000e+07    116.000000      4.000000   
max        50.000000  7.630000e+08    900.000000     19.000000   

        ceiling_height  floors_total   living_area         floor is_apartment  \
count     14504.000000  23613.000000  21796.000000  23699.000000         2775   
unique             NaN           NaN           NaN           NaN            2   
top                NaN           NaN           NaN           NaN        False   
freq               NaN           NaN           NaN           NaN         2725   
mean          2.771499     10.673824     34.457852      5.892358          NaN   
std           1.261056      6.597173     22.030445      4.885249          NaN   
min           1.000000      1.000000      2.000000      1.000000          NaN   
3%            2.500000      3.000000     14.700000      1.000000          NaN   
25%           2.520000      5.000000     18.600000      2.000000          NaN   
50%           2.650000      9.000000     30.000000      4.000000          NaN   
75%           2.800000     16.000000     42.300000      8.000000          NaN   
95%           3.300000     25.000000     69.000000     16.000000          NaN   
max         100.000000     60.000000    409.700000     33.000000          NaN   

       studio open_plan  kitchen_area       balcony  airports_nearest  \
count   23699     23699  21421.000000  12180.000000      18157.000000   
unique      2         2           NaN           NaN               NaN   
top     False     False           NaN           NaN               NaN   
freq    23550     23632           NaN           NaN               NaN   
mean      NaN       NaN     10.569807      1.150082      28793.672193   
std       NaN       NaN      5.905438      1.071300      12630.880622   
min       NaN       NaN      1.300000      0.000000          0.000000   
3%        NaN       NaN      5.200000      0.000000      10663.000000   
25%       NaN       NaN      7.000000      0.000000      18585.000000   
50%       NaN       NaN      9.100000      1.000000      26726.000000   
75%       NaN       NaN     12.000000      2.000000      37273.000000   
95%       NaN       NaN     20.000000      2.000000      51340.000000   
max       NaN       NaN    112.000000      5.000000      84869.000000   

        cityCenters_nearest  parks_around3000  parks_nearest  \
count          18180.000000      18181.000000    8079.000000   
unique                  NaN               NaN            NaN   
top                     NaN               NaN            NaN   
freq                    NaN               NaN            NaN   
mean           14191.277833          0.611408     490.804555   
std             8608.386210          0.802074     342.317995   
min              181.000000          0.000000       1.000000   
3%              2617.220000          0.000000      58.000000   
25%             9238.000000          0.000000     288.000000   
50%            13098.500000          0.000000     455.000000   
75%            16293.000000          1.000000     612.000000   
95%            31671.600000          2.000000     968.000000   
max            65968.000000          3.000000    3190.000000   

        ponds_around3000  po

#### Дублирующиеся записи

In [ ]:
df.duplicated()

0        False
1        False
2        False
3        False
4        False
         ...  
23694    False
23695    False
23696    False
23697    False
23698    False
Length: 23699, dtype: bool

#### Пропущенные значения

In [ ]:
df.isna().sum()

total_images                0
last_price                  0
total_area                  0
first_day_exposition        0
rooms                       0
ceiling_height           9195
floors_total               86
living_area              1903
floor                       0
is_apartment            20924
studio                      0
open_plan                   0
kitchen_area             2278
balcony                 11519
locality_name              49
airports_nearest         5542
cityCenters_nearest      5519
parks_around3000         5518
parks_nearest           15620
ponds_around3000         5518
ponds_nearest           14589
days_exposition          3181
dtype: int64

|Признак|Комментарий|
|--:|:--|
|`days_exposition`|<s>Привести тип к `float`, исследовать аномалии, пропуски</s>|


#### Доступ к данным

Доступ: 
- столбцам;
- строкам;
- срезы.

`Через точку` (атрибутивный) и через `указание имени`.

##### Доступ к столбцам 

In [ ]:
df['total_area'];

In [ ]:
lst_col = ['total_images', 'last_price', 'total_area']
df[lst_col]

total_images  last_price  total_area
0                20    13000000       108.0
1                 7     3350000        40.4
2                10     5196000        56.0
3                 0    64900000       159.0
4                 2    10000000       100.0
...             ...         ...         ...
23694             9     9700000      133.81
23695            14     3100000        59.0
23696            18     2500000        56.7
23697            13    11475000       76.75
23698             4     1350000        32.3

[23699 rows x 3 columns]

##### Доступ к строкам

In [ ]:
df[1:4:2]

total_images  last_price  total_area first_day_exposition  rooms  \
1             7     3350000        40.4  2018-12-04T00:00:00      1   
3             0    64900000       159.0  2015-07-24T00:00:00      3   

   ceiling_height  floors_total  living_area  floor  is_apartment  ...  \
1            <NA>            11         18.6      1          <NA>  ...   
3            <NA>            14         <NA>      9          <NA>  ...   

   kitchen_area  balcony    locality_name  airports_nearest  \
1          11.0        2   посёлок Шушары             12817   
3          <NA>        0  Санкт-Петербург             28098   

  cityCenters_nearest  parks_around3000  parks_nearest  ponds_around3000  \
1               18603                 0           <NA>                 0   
3                6800                 2             84                 3   

   ponds_nearest  days_exposition  
1           <NA>             81.0  
3            234            424.0  

[2 rows x 22 columns]

total_images  last_price  total_area first_day_exposition  rooms  \
1             7   3350000.0        40.4  2018-12-04T00:00:00      1   
2            10   5196000.0        56.0  2015-08-20T00:00:00      2   

   ceiling_height  floors_total  living_area  floor is_apartment  ...  \
1             NaN          11.0         18.6      1          NaN  ...   
2             NaN           5.0         34.3      4          NaN  ...   

   kitchen_area  balcony    locality_name  airports_nearest  \
1          11.0      2.0   посёлок Шушары           12817.0   
2           8.3      0.0  Санкт-Петербург           21741.0   

  cityCenters_nearest  parks_around3000  parks_nearest  ponds_around3000  \
1             18603.0               0.0            NaN               0.0   
2             13933.0               1.0           90.0               2.0   

   ponds_nearest  days_exposition  
1            NaN             81.0  
2          574.0            558.0  

[2 rows x 22 columns]

##### [Работа с `.loc[]` `.iloc[]`](https://towardsdatascience.com/how-to-use-loc-and-iloc-for-selecting-data-in-pandas-bd09cb4c3d79)

In [ ]:
dict_to_df = {'student_fio': ['Иванов', 'Петров', 'Сидорова', 'Волкова', 'Сотников'],
              'gender' : ['M', 'M', 'F', 'F', 'M'],
              'age' : [23, 23, 24, 25, 36],
              'curs' : [1,2,2,4,4]}

df = pd.DataFrame(dict_to_df, index = ['st1', 'st2', 'st3', 'st4', 'st5'])
df

student_fio gender  age  curs
st1      Иванов      M   23     1
st2      Петров      M   23     2
st3    Сидорова      F   24     2
st4     Волкова      F   25     4
st5    Сотников      M   36     4

In [ ]:
df.loc['st1']

pandas.core.series.Series

In [ ]:
df.loc[['st1', 'st2']]

pandas.core.frame.DataFrame

In [ ]:
df.loc[:,['student_fio', 'gender']]

student_fio gender
st1      Иванов      M
st2      Петров      M
st3    Сидорова      F
st4     Волкова      F
st5    Сотников      M

In [ ]:
df.loc[['st1', 'st2'], :]

student_fio gender  age  curs
st1      Иванов      M   23     1
st2      Петров      M   23     2

In [ ]:
df.loc[['st1', 'st2'],['student_fio', 'gender']]

student_fio gender
st1      Иванов      M
st2      Петров      M

In [ ]:
df.loc['st1':'st5', ['student_fio', 'gender']]

student_fio gender
st1      Иванов      M
st2      Петров      M
st3    Сидорова      F
st4     Волкова      F
st5    Сотников      M

In [ ]:
df.loc[[True, True, True, False, False], ['student_fio', 'gender']]

student_fio gender
st1      Иванов      M
st2      Петров      M
st3    Сидорова      F

In [ ]:
df.iloc[[1,3,4]]

student_fio gender  age  curs
st2      Петров      M   23     2
st4     Волкова      F   25     4
st5    Сотников      M   36     4

In [ ]:
df.iloc[[1,3,4], [1,2]]

gender  age
st2      M   23
st4      F   25
st5      M   36

In [ ]:
df.iloc[:, [True, True, False, False]]

student_fio gender
st2      Петров      M
st4     Волкова      F
st5    Сотников      M

Замечание: С помощью цепочки методов `.index.get_loc('здесь_метка_индекса')` можно получить номер позиции метки индекса и воспользоваться методом `.iloc[]`  

##### Доступ к скалярному значению по метке `.at[]` и по позиции `iat[]`

In [ ]:
df.at['st4', 'student_fio'] = "Зайцев"

In [ ]:
df

student_fio gender  age  curs
st1      Иванов      M   23     1
st2      Петров      M   23     2
st3    Сидорова      F   24     2
st4      Зайцев      F   25     4
st5    Сотников      M   36     4

In [ ]:
df.iat[3, 0]

'Волкова'

#### Фильтрация данных

Логический отбор

оператор `isin()`

метод `.query()` про inplace = True

In [ ]:
df[(df.ceiling_height >= 2.2) & 
       (df.ceiling_height <= 4) & 
       (df.rooms == 2)].shape[0]

4826

In [ ]:
grade = [2]

In [ ]:
df[df.ceiling_height.isin([2, 2.5])]

total_images  last_price  total_area first_day_exposition  rooms  \
9                18   5400000.0        61.0  2017-02-26T00:00:00      3   
18                8   3250000.0        31.0  2017-01-27T00:00:00      1   
37               10   1990000.0        45.8  2017-10-28T00:00:00      2   
49                1   3050000.0        30.8  2018-11-22T00:00:00      1   
64               13   3100000.0        44.3  2017-11-07T00:00:00      2   
...             ...         ...         ...                  ...    ...   
23681            13   5250000.0        43.0  2018-05-11T00:00:00      2   
23682            11   2350000.0        35.6  2016-06-23T00:00:00      1   
23683            16   2100000.0        62.8  2018-09-18T00:00:00      4   
23685            15   2490000.0        31.0  2019-01-24T00:00:00      1   
23698             4   1350000.0        32.3  2017-07-21T00:00:00      1   

       ceiling_height  floors_total  living_area  floor is_apartment  ...  \
9                 2.5           9.0         43.6      7          NaN  ...   
18                2.5           5.0         19.4      2          NaN  ...   
37                2.5           5.0          NaN      1          NaN  ...   
49                2.5           9.0         18.0      7          NaN  ...   
64                2.5           5.0         29.0      5          NaN  ...   
...               ...           ...          ...    ...          ...  ...   
23681             2.5           5.0         29.0      2          NaN  ...   
23682             2.5           5.0         17.0      4        False  ...   
23683             2.5           5.0         45.5      3          NaN  ...   
23685             2.5           5.0         17.3      5          NaN  ...   
23698             2.5           5.0         12.3      1          NaN  ...   

       kitchen_area  balcony                        locality_name  \
9               6.5      2.0                      Санкт-Петербург   
18              5.6      1.0                      Санкт-Петербург   
37              NaN      NaN  поселок городского типа Красный Бор   
49              6.0      NaN                      Санкт-Петербург   
64              5.3      NaN                              Гатчина   
...             ...      ...                                  ...   
23681           6.0      NaN                      Санкт-Петербург   
23682           8.3      2.0                               Выборг   
23683           5.5      NaN                 посёлок Дзержинского   
23685           5.6      1.0                            Ломоносов   
23698           9.0      NaN                  поселок Новый Учхоз   

       airports_nearest cityCenters_nearest  parks_around3000  parks_nearest  \
9               50898.0             15008.0               0.0            NaN   
18              44060.0             10842.0               1.0          759.0   
37                  NaN                 NaN               NaN            NaN   
49              45341.0             16603.0               1.0          675.0   
64                  NaN                 NaN               NaN            NaN   
...                 ...                 ...               ...            ...   
23681           28323.0              7262.0               1.0          497.0   
23682               NaN                 NaN               NaN            NaN   
23683               NaN                 NaN               NaN            NaN   
23685           48393.0             51818.0               0.0            NaN   
23698               NaN                 NaN               NaN            NaN   

       ponds_around3000  ponds_nearest  days_exposition  
9                   0.0            NaN            289.0  
18                  0.0            NaN            168.0  
37                  NaN            NaN            196.0  
49                  0.0            NaN              NaN  
64                  NaN            NaN            143.0  
...                 ...            ...     

In [ ]:
df[df['curs'].isin([1, 2])]

student_fio gender  age  curs
st1      Иванов      M   23     1
st2      Петров      M   23     2
st3    Сидорова      F   24     2

In [ ]:
df.query("gender == 'M' and curs == 2")

student_fio gender  age  curs
st2      Петров      M   23     2

##### Отбор строк и столбцов через фильтрацию

In [ ]:
df.loc[((df.curs.isin([1,2])) &
        (df.gender == 'F')), 'student_fio']

st3    Сидорова
Name: student_fio, dtype: object

##### Модификация датафрейма

In [ ]:
df.loc[~((df.curs.isin([1,2])) & (df.gender == 'F')), 'new_colun']=1
df


student_fio gender  age  curs  new_colun
st1      Иванов      M   23     1        1.0
st2      Петров      M   23     2        1.0
st3    Сидорова      F   24     2        NaN
st4     Волкова      F   25     4        1.0
st5    Сотников      M   36     4        1.0

In [ ]:
df.loc['st3', 'age'] = 123
df


student_fio gender  age  curs  new_colun
st1      Иванов      M   23     1        1.0
st2      Петров      M   23     2        1.0
st3    Сидорова      F  123     2        NaN
st4     Волкова      F   25     4        1.0
st5    Сотников      M   36     4        1.0

In [ ]:
df[df.age == 23].age

st1    23
st2    23
Name: age, dtype: int64

In [ ]:
df

student_fio gender  age  curs
st1      Иванов      M   23     1
st2      Петров      M   23     2
st3    Сидорова      F  123     2
st4     Волкова      F   25     4
st5    Сотников      M   36     4

In [ ]:
df.groupby(['gender'])['age', 'curs'].agg(['min', 'max', 'mean'])

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  """Entry point for launching an IPython kernel.


age                 curs              
       min  max       mean  min max      mean
gender                                       
F       25  123  74.000000    2   4  3.000000
M       23   36  27.333333    1   4  2.333333

In [ ]:
df

student_fio gender  age  curs
st1      Иванов      M   23     1
st2      Петров      M   23     2
st3    Сидорова      F   24     2
st4     Волкова      F   25     4
st5    Сотников      M   36     4

In [ ]:
df.groupby(['curs', 'gender']).age.agg(['min', 'max', 'mean'])

min  max  mean
curs gender                
1    M        23   23  23.0
2    F        24   24  24.0
     M        23   23  23.0
4    F        25   25  25.0
     M        36   36  36.0

## Литература

1. [Маккинли У. Python и анализ данных. $-$ 2-е изд. $-$ М.: ДМК Пресс, 2020. $–$ 540 с.](https://dmkpress.com/catalog/computer/programming/python/978-5-97060-590-5/)
2. [Хейдт М. Изучаем `pandas` / М. Хейдт, А.В. Груздев. $-$ 2-е изд., дополненное. $-$ М.: ДМК Пресс, 2019. $–$ 700 с.  ](https://dmkpress.com/catalog/computer/programming/python/978-5-97060-670-4/)
3. [Абдрахманов М.И. Devpractice Team. Pandas. Работа с данными. 2-е изд. - devpractice.ru. 2020. - 170 с.](https://devpractice.ru/kniga-pandas/)
4. [Прохоренок Н.А. Python 3 и PyQt 5. Разработка приложений. — 2-е изд., перераб. и доп. / Н.А. Прохоренок, В.А. Дронов. — СПб.: БХВ-Петербург, 2018. — 832 с.](https://bhv.ru/product/python-3-i-pyqt-5-razrabotka-prilozhenij-2-izd/)